### DISCUSSION TITLE
Fetch by "key" in array

### DISCUSSION THREAD

**Discussion by richcorbs** [deta/cloud-docs#230] [Help] [answered]
I'm trying to do the equivalent of an "in" query in Base like so:

```
db.projects.fetch({ "key": ["key-1", "key-2", "key-3"] })
```
It doesn't seem to work.

The documentation makes it look like this pattern works. But maybe it doesn't work for the "key" field?

```
{"fav_numbers": [2, 4, 8]}
```

Is there another way? I'd rather not have to make multiple queries (one per array element) to get what I'm after.

**Comment by jnsougata** [comment, marked as answer]
[ ❌ ] `db.projects.fetch([{ "key": "key-1"}, { "key": "key-2"}.....{ "key": "key-N"}])`
for reference check: https://docs.deta.sh/docs/base/queries/#equal

**Reply by richcorbs** [reply]
Thanks for the response!

I had referenced that section in the documentation and it was not even remotely apparent to me that the array of objects syntax was even a possibility.

It seems that that syntax will work for what I need.

**Reply by richcorbs** [reply]
I can't seem to get this working. Are we sure this pattern works?

Passing an array of objects to the fetch returns nothing. It's like it errors out.

If I pass 1 object in the array it works.

**Reply by jnsougata** [reply]
you should check `querying-keys` https://docs.deta.sh/docs/base/queries/#querying-keys

**Reply by richcorbs** [reply]
Yep, I've seen that, tried to replicate it, and it doesn't seem to be working with the javascript SDK.

```
db.projects.fetch({ [{"key?pfx":"key-1"}, {"key?pfx": "key-2"}] })
db.projects.fetch({ [{"key":"key-1"}, {"key": "key-2"}] })
db.projects.fetch([{"key":"key-1"}, {"key": "key-2"}])
```
Nothing works.

**Reply by jnsougata** [reply]
![image](https://user-images.githubusercontent.com/53375272/176597018-1a89bfcb-7b8e-442c-92f7-09b3c2a6772a.png)
**you literally misread the thing. these are an example of `invalid queries` which you are trying to replicate.**
Primary key query doesn't support `AND` `OR` operations therefore you can't do query it this way

**Reply by richcorbs** [reply]
I'm confused.

Your suggestion is one of those invalid patterns.

Maybe those queries are invalid because they don't work with a key of `key`?

**Reply by jnsougata** [reply]
yes, previously I did the same mistake. the query we are trying to make is invalid. rather you can write a func and fetch those keys from base asynchronously

**Reply by abdelhai** [reply]
@richcorbs both of them are invalid

**Reply by richcorbs** [reply]
So at this point the only option is to loop through the ids and get/fetch them one at a time?

If I were querying by some other key the OR query would work. Correct?

**Reply by abdelhai** [reply]
try to figure out a way to query based on a different attribute other than the `key`. keys are special but limited for querying.
for keys, you still can query based on key prefixes and ranges, etc

**Reply by richcorbs** [reply]
I guess I could denormalize the key into a queryable key field. Looping seems ok for now.

**Comment by amjed-ali-k** [comment]
To avoid multiple queries, you can create a new duplicate field for `key` on each item. 
```json
{
"key" : "0c9f904ffb9540f7a1539add854b7bc3",
"id": "0c9f904ffb9540f7a1539add854b7bc3",
"value": "Sample"
}
```

and query them by the duplicate field.

```python
  deta.Base("test").fetch([ {"id": "0c9f904ffb9540f7a1539add854b7bc3"}, {"id": "093889577a4b4955be79e8fde331720e"}]).items
```
